streams/shared/src/main/scala/zio/stream/ZPipeline.scala

```diff
@@ -1798,6 +1817,52 @@ object ZPipeline extends ZPipelinePlatformSpecificConstructors {
     new ZPipeline(loop(Chunk.ChunkIterator.empty, 0))
   }
 
+  /**
+   * Creates a pipeline that maps over elements of the stream with the specified
+   * effectful function.
+   *
+   * Unlike `mapZIO` processing is done chunk by chunk. This means that
+   * `mapZIOChunked` provides weaker guarantees than `mapZIO`. While
+   * `stream.mapZIO(f).mapZIO(g)` is guaranteed to be equivalent to
+   * `stream.mapZIO(x => f(x).flatMap(g))`, the same is not true for
+   * `mapZIOChunked`. For example, `mapZIO` guarantees that the first element of
+   * a stream will first be processed with `f` and then `g` before the second
+   * element is processed with `f`. `mapZIOChunked` may process the first two
+   * elements with `f` and only then move on to process the first element with
+   * `g`.
+   */
+  def mapZIOChunked[Env, Err, In, Out](
+    f: In => ZIO[Env, Err, Out]
+  )(implicit trace: Trace): ZPipeline[Env, Err, In, Out] = {
+    def loop(
+      chunkIterator: Chunk.ChunkIterator[In],
+      index: Int,
+      builder: ChunkBuilder[Out]
+    ): ZChannel[Env, Err, Chunk[In], Any, Err, Chunk[Out], Any] =
+      if (chunkIterator.hasNextAt(index)) {
+        ZChannel.unwrap {
+          val a = chunkIterator.nextAt(index)
+          f(a).foldCause(
+            cause => ZChannel.write(builder.result()) *> ZChannel.refailCause(cause),
```

Comment:
This needs an index check, if this is the first value in the chunk we should skip writing an empty chunk and immediately write the error

Reply:
Why? At this point the stream is failed anyway, so it's not worth optimizing this.

Reply:
I agree. Emitting empty chunks doesn't make sense. I'll update it.

Reply:
Fixed in https://github.com/zio/zio/pull/9697/commits/cf506a767a0f2eeec78529ba40b25632340afd29